### PR TITLE
fix issue 5556: arange floating point error

### PR DESCRIPTION
**Pull request by Officium**
One can use `a / c - b / c` to reduce floating point error of `(a - b) / c`

### CODE REVIEW

**Comment by Officium** [comment]
[Issue5556](https://github.com/pytorch/pytorch/issues/5556)

**Comment by ezyang** [comment]
Thank you! Is there a test we can add for this case?

**Comment by ezyang** [comment]
Actually, it looks like this intersects with #5600

**Comment by Officium** [comment]
We can verify this in simple cpp.
```cpp
#include <iostream>

int main(){
    ptrdiff_t size = (ptrdiff_t) (((0.45 - 0.1) / 0.05) + 1);
    std::cout << size << std::endl;
    size  = (ptrdiff_t) ((0.45 / 0.05 - 0.1 / 0.05) + 1);
    std::cout << size << std::endl;
    return 0;
}
```
And in pytorch, run cmd
```shell
python -c "import torch;print torch.arange(0.1, 0.50, 0.05)"
```

**Comment by zou3519** [comment]
@Officium Is there a mathematical reason why ` a / c - b / c` has less floating point error than `(a - b) / c`?

**Comment by ezyang** [comment]
@Officium OK. Can we add this as a test to `test/test_torch.py` using `assertEqual` with a suitably small `prec`?

**Comment by Officium** [comment]
@ezyang Ok. I will add a test and fix the same issue in /aten/src/TH/generic/THTensorMath.c.

**Comment by vadimkantorov** [comment]
Here is NumPy impl: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/ctors.c#L3040

**Comment by Officium** [comment]
@vadimkantorov I trace these numpy codes, and only find it use function "npy_ceil" which I can't find the implementation.

**Comment by Officium** [comment]
@zou3519 I just remind there is a similar way to reduce computational error in the field of numberical calculation.

**Comment by vadimkantorov** [comment]
```
_calc_length:
*next = PyNumber_Subtract(stop, start);
val = PyNumber_TrueDivide(*next, step);
val = _arange_safe_ceil_to_intp(val)

_arange_safe_ceil_to_intp(double value):
ivalue = npy_ceil(value);
return (npy_intp)ivalue;

```

NumPy has two different impls of PyArray_Arange: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/ctors.c#L2968 and https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/ctors.c#L3111

npy_ceil is auto-generated here: https://github.com/numpy/numpy/blob/d407f24cb5ad85850a60b1be3dade3305ea30c98/numpy/core/src/npymath/npy_math_internal.h.src#L359

I hope we are doing operations in double precision here.

**Comment by apaszke** [comment]
@zou3519 I think this is a trick to decrease the absolute roundoff error (the relative error stays constant).

**Comment by li-roy** [comment]
I think we need to address the floating point issue in the arange function as well. I think the fmod call is a source of error.

**Comment by Officium** [comment]
@zou3519 @li-roy @ezyang @apaszke @vadimkantorov 
I get an example of this trick.
If we only have 3 significant digits for any number, and we want to calculate `8 - sqrt(63)`. Then we will get `8 - sqrt(63) = 8 - 7.94 = 0.06`. We can just transform the calculation format to get higher accuracy `8 - sqrt(63) = 1 / (8 + sqrt(63)) = 1 / (8 + 7.94) = 1 / 15.9 = 0.0629`.
In this example, we can see that subtraction of two close number will loss accuracy because of the store space. In the commit of this pull request, the loss is transformed from `start` and `stop` to `start / step` and `stop / step` where the distance between the latter pair is bigger in `arange` setting. 
I saw #5600 and I think it is important to make consistent with `numpy.arange`. In the other hand, we should get more accuracy result as possible as we can.

**Comment by apaszke** [comment]
@Officium yes, subtraction is the operation with worst numerical properties here, but your example doesn't hold, because we're unable to replace it with addition. What really happens is that you still get the loss of precision, but the *absolute* loss you see is smaller if you do it on smaller numbers (relative errors stay the same).

**Comment by Officium** [comment]
@apaszke I agree with you. Relative errors stay the same. My example is not very suitable. I want to show the relationship of absolute loss and the subtraction distance. Computational error can't be reduced to zero. We are only able to reduce both absolute error and relative error as many as possible.
